spec.bs

```diff
@@ -201,68 +210,264 @@ This document introduces a new storage API that is intentionally not partitioned
 
+  Note: The website that serves the module script should be aware of the implication of CORS: when the the module script's [=/URL=]'s [=url/origin=] and the worklet's creator window's origin is not [=same origin=], and by granting the module script resource via CORS, it will also grant the worklet's creation and subsequent operations on the worklet, under module script's [=/URL=]'s [=url/origin=]. For example, the worklet's creator context could poison and use up the [=SharedStorageWorklet/worklet origin=]'s [=remaining navigation budget=].
+
+  ## Run Operation Methods on {{SharedStorageWorklet}} ## {#run-op-shared-storage-worklet}
+
+  <div algorithm>
+    To <dfn>get the select-url result index</dfn>, given {{SharedStorageWorklet}} |worklet|, {{DOMString}} |operationName|, [=/list=] of {{SharedStorageUrlWithMetadata}}s |urlList|, and {{SharedStorageRunOperationMethodOptions}} |options|:
+
+    1. Let |promise| be a new [=promise=].
+    1. Let |window| be |worklet|'s [=relevant settings object=].
+    1. [=Assert=]: |window| is a {{Window}}.
+    1. If |window|'s [=Window/browsing context=] is null, then return a [=promise rejected=] with a {{TypeError}}.
+    1. If |window|'s [=associated document=] is not [=fully active=], return a [=promise rejected=] with a {{TypeError}}.
+    1. Return |promise|, and immediately [=obtain a worklet agent=] given |window| and run the rest of these steps in that agent:
+        1. Let |index| be [=default index=].
+        1. [=Assert=]: |worklet|'s [=global scopes=]'s [=list/size=] is 1.
+        1. [=Assert=]: |worklet|'s [=module map=] is not [=map/empty=].
+        1. Let |globalScope| be [=this=]'s [=global scopes=][0].
+        1. Let |operationMap| be |globalScope|'s [=SharedStorageWorkletGlobalScope/operation map=].
+        1. If |operationMap| does not [=map/contain=] |operationName|, then [=queue a global task=] on the [=DOM manipulation task source=], given |window|, to [=reject=] |promise| with a {{TypeError}}, and abort these steps.
+
+            Note: This could happen if {{SharedStorageWorkletGlobalScope/register()}} was never called with |operationName|.
+
+        1. Let |operation| be |operationMap|[|operationName|].
+        1. Let |argumentsList| be the [=/list=] « |urlList| ».
+        1. If |options| [=map/contains=] |data|, [=list/append=] |data| to |argumentsList|.
+        1. Let |operationResult| be the result of [=invoking=] |operation| with |argumentsList|.
+        1. If [=an exception was thrown=], then [=queue a global task=] on the [=DOM manipulation task source=], given |window|, to [=reject=] |promise| with a {{TypeError}}.
+
+            Note: This indicates that |operationCtor|'s run() method encounters an error, where |operationCtor| is the parameter in {{SharedStorageWorkletGlobalScope/register()}}.
+
+        1. Otherwise:
+            1. Set |index| to |operationResult|, [=converted to a JavaScript value=] of type {{unsigned long}}.
```

Comment:
It is supposed to be a `run()` function inside the class (corresponding to the constructor). For example, inside the module script, it'd do
```
class SelectURLOperation {
  async run(data, urls) {
    return 3;
  }
}

register('select-url-for-experiment', SelectURLOperation);
```

In the [Chrome code](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/modules/shared_storage/shared_storage_worklet_global_scope.cc?q=%22retriever.GetMethodOrThrow%22), we don't define any IDL, but rather we "manually" retrieve the the run() method. In the [spec](https://wicg.github.io/shared-storage/#dom-sharedstorageworkletglobalscope-register ), we have _"Let run be the result of running [GetMethod()](https://tc39.es/ecma262/#sec-getmethod) with prototype and "run"._

I believe something is still missing, but I don't know how to specify this properly. Should we define the type somewhere (via IDL)? Should we not do [=invoke=] here, but do something else in order to "call this function"?

Reply:
Updated following the worklet example. Also removed the definition and references for those "Operation" types.

But the original concern remains: At the [=invoke=] time, as you said, it already returns an IDL type. But how it's specified to be {{unsigned long}} / how do I convert it to that type?

Reply:
> But the original concern remains: At the [=invoke=] time, as you said, it already returns an IDL type. But how it's specified to be {{unsigned long}} / how do I convert it to that type?

You'll make an IDL callback type like https://wicg.github.io/observable/#callbackdef-predicate, but instead of `bool` as the return type, you want `unsigned long` I think. You can choose whatever arguments as well. Then instead of:

> Let callback be the result of [converting](https://webidl.spec.whatwg.org/#es-type-mapping) function to a Web IDL [Function](https://webidl.spec.whatwg.org/#common-Function) instance.

You'll convert to a Web IDL "your callback type here", type.

Reply:
Got it. Done. (Introduced callback `RunFunctionForSharedStorageSelectURLOperation` type, and updated the usage site.)